### PR TITLE
refactor: parse once from base command

### PR DESCRIPTION
**Pull request by john-u**
Implements https://github.com/oclif/oclif.github.io/pull/142

Highlights
- Same performance as before
- Maintains type safety of flags in child commands
- Adds type safety to base-level command flags

## Checklist

- [x] I have read the **[CONTRIBUTING](../CONTRIBUTING.md)** document
- [x] Any required documentation has been added
- [x] My code follows the code style of this project (`npm run lint` produces no warnings/errors)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: packages/cli/src/commands/apps.ts [hunk 1]
```diff
@@ -35,27 +35,27 @@ export default class AppsCommand extends APICommand {
...
+			if (this.flags.type) {
+				appListOptions.appType = AppType[this.flags.type as keyof typeof AppType]
 			}
-			if (flags.classification) {
-				appListOptions.classification = flags.classification.map(it => AppClassification[it as keyof typeof AppClassification])
+
+			if (this.flags.classification) {
+				// tradeoff for not parsing in here is that TS cannot infer typeof this.constructor.flags
+				const classifications = this.flags.classification as string[]
+				appListOptions.classification = classifications.map(classification => AppClassification[classification as keyof typeof AppClassification])
```
Comment (john-u) [inline packages/cli/src/commands/apps.ts:58]: I tried a number of things both here and in out base command, but couldn't get flags to infer the static flags types without explicit casting.

Comment (john-u) [inline packages/cli/src/commands/apps.ts:58]: I tried a number of things both here and in out base command, but couldn't get flags to infer the static flags types without explicit casting.

File: packages/cli/src/commands/capabilities.ts [hunk 2]
```diff
@@ -131,9 +135,9 @@ export interface CapabilityId {
 
 export type CapabilitySummaryWithNamespace = CapabilitySummary & { namespace: string }
 
-export function buildListTableOutput(this: APIOrganizationCommand, capabilities: CapabilitySummaryWithNamespace[]): string {
-	return this.tableGenerator.buildTableFromList(capabilities, ['id', 'version', 'status'])
-}
+// export function buildListTableOutput(this: APIOrganizationCommand, capabilities: CapabilitySummaryWithNamespace[]): string {
+// 	return this.tableGenerator.buildTableFromList(capabilities, ['id', 'version', 'status'])
+// }
```
Comment (rossiam) [inline packages/cli/src/commands/capabilities.ts:140]: I'm betting you meant to remove this completely after compiling to make sure it wasn't used.

File: packages/lib/src/smartthings-command.ts [hunk 3]
```diff
@@ -145,70 +177,37 @@ export abstract class SmartThingsCommand extends Command implements SmartThingsC
...
 		const compact = this.flags.expanded
 			? false
 			: (this.flags.compact ? true : this.booleanConfigValue('compactTableOutput', true))
+
 		this._tableGenerator = new DefaultTableGenerator(compact)
 	}
 
 	/**
-	 * This method should be called when the user has decided to not complete a command.
+	 * @deprecated use Command.exit() instead
```
Comment (rossiam) [inline packages/lib/src/smartthings-command.ts:208]: Why not fix this method to use `Command.exit` instead? It also logs a message for the user which `Command.exit` doesn't do and is something we usually want to do in this case.

**Review by rossiam — APPROVED** [review]

File: packages/cli/src/commands/capabilities.ts (hunk 2 above)
Comment (john-u) [reply packages/cli/src/commands/capabilities.ts:140]: I sure did. Thanks.

**Pull request merged**